cmd/package.go

```diff
@@ -60,83 +59,107 @@ func newPlistData(title, exe, packageID, version, author string) *plistData {
 	}
 }
 
-type windowsIconHeader struct {
+type windowsIcoHeader struct {
 	_          uint16
 	imageType  uint16
 	imageCount uint16
-	width      uint8
-	height     uint8
-	colours    uint8
-	_          uint8
-	planes     uint16
-	bpp        uint16
-	size       uint32
-	offset     uint32
+}
+
+type windowsIcoDescriptor struct {
+	width   uint8
+	height  uint8
+	colours uint8
+	_       uint8
+	planes  uint16
+	bpp     uint16
+	size    uint32
+	offset  uint32
+}
+
+type windowsIcoContainer struct {
+	Header windowsIcoDescriptor
+	Data   []byte
 }
 
 func generateWindowsIcon(pngFilename string, iconfile string) error {
-	header := &windowsIconHeader{
-		imageType:  1,
-		imageCount: 1,
-		bpp:        32,
-		planes:     1,
-		offset:     22,
-		width:      0,
-		height:     0,
-	}
+	sizes := []int{256, 128, 64, 48, 32, 16}
 
-	// Load png
 	pngfile, err := os.Open(pngFilename)
 	if err != nil {
 		return err
 	}
 	defer pngfile.Close()
 
-	// Decode to internal image
 	pngdata, err := png.Decode(pngfile)
 	if err != nil {
 		return err
 	}
 
-	// Scale to 256*256
-	rect := image.Rect(0, 0, 255, 255)
-	rawdata := image.NewRGBA(rect)
-	scale := draw.ApproxBiLinear
-	scale.Scale(rawdata, rect, pngdata, pngdata.Bounds(), draw.Over, nil)
-
-	// Convert back to PNG
-	icondata := new(bytes.Buffer)
-	writer := bufio.NewWriter(icondata)
-	err = png.Encode(writer, rawdata)
+	outfile, err := os.Create(iconfile)
 	if err != nil {
 		return err
 	}
-	err = writer.Flush()
+	defer outfile.Close()
+
+	ico := windowsIcoHeader{
+		imageType:  1,
+		imageCount: uint16(len(sizes)),
+	}
+	err = binary.Write(outfile, binary.LittleEndian, ico)
```

Comment:
If we write out here and error later, then we'll have corrupt .ico files written. Suggest moving this to the end once we know everything has been generated.

Reply:
I got the point. we should just delete the corrupted file (if any) once this method exits with an error.